storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/RedisOnlineRetrieverV2.java

```diff
@@ -0,0 +1,195 @@
+    entityNames.sort(String::compareTo);
+
+    for (String entityName : entityNames) {
+      builder.addEntityNames(entityName);
+      builder.addEntityValues(fieldsMap.get(entityName));
+    }
+    return builder.build();
+  }
+
+  private List<List<Optional<Feature>>> getFeaturesFromRedis(
+      List<RedisKeyV2> redisKeys, List<FeatureReferenceV2> featureReferences) {
+    List<List<Optional<Feature>>> features = new ArrayList<>();
+    // To decode bytes back to Feature Reference
+    Map<String, FeatureReferenceV2> byteToFeatureReferenceMap = new HashMap<>();
+    // To check whether redis ValueK is a timestamp field
+    Map<String, Boolean> isTimestampMap = new HashMap<>();
+
+    // Serialize using proto
+    List<byte[]> binaryRedisKeys =
+        redisKeys.stream().map(redisKey -> redisKey.toByteArray()).collect(Collectors.toList());
+
+    try {
+      List<List<byte[]>> featureReferenceWithTsByteArrays = new ArrayList<>();
+      featureReferences.stream()
+          .forEach(
+              featureReference -> {
+                List<byte[]> currentFeatureWithTsByteArrays = new ArrayList<>();
+
+                // eg. murmur(<featuretable_name:feature_name>)
+                String delimitedFeatureReference =
+                    featureReference.getFeatureTable() + ":" + featureReference.getName();
+                byte[] featureReferenceBytes =
+                    Hashing.murmur3_32()
+                        .hashString(delimitedFeatureReference, StandardCharsets.UTF_8)
+                        .asBytes();
+                currentFeatureWithTsByteArrays.add(featureReferenceBytes);
+                isTimestampMap.put(Arrays.toString(featureReferenceBytes), false);
+                byteToFeatureReferenceMap.put(featureReferenceBytes.toString(), featureReference);
+
+                // eg. <_ts:featuretable_name>
+                String timestampFeatureTableReference =
+                    timestampPrefix + ":" + featureReference.getFeatureTable();
+                byte[] featureTableTsBytes = timestampFeatureTableReference.getBytes();
+                isTimestampMap.put(Arrays.toString(featureTableTsBytes), true);
+                currentFeatureWithTsByteArrays.add(featureTableTsBytes);
+
+                featureReferenceWithTsByteArrays.add(currentFeatureWithTsByteArrays);
+              });
+
+      // Access redis keys and extract features
+      for (byte[] binaryRedisKey : binaryRedisKeys) {
+        List<Optional<Feature>> curRedisKeyFeatures = new ArrayList<>();
+        // Loop according to each FeatureReferenceV2 bytes
+        for (List<byte[]> currentFeatureReferenceWithTsByteArray :
+            featureReferenceWithTsByteArrays) {
+          FeatureReferenceV2 featureReference = null;
+          ValueProto.Value featureValue = null;
+          Timestamp eventTimestamp = null;
+
+          // Always 2 fields (i.e feature and timestamp)
+          List<KeyValue<byte[], byte[]>> redisValuesList =
```

Comment:
so you creating request to redis to read one feature & timestamp
instead it should be
`hmget key (all_features + all_timestamps per key)`

Reply:
Also for optimization purposes it's better to use redis pipelines here.
First all request sent to redis,
`List<Future<Result>> results = keys.map(pipeline.hgmet(...))`
 then all requests processed.
`results.foreach(...)`